src/rsa/blinding.rs

```diff
@@ -12,52 +12,127 @@
+                        blinding_factor_inv: blinding_factor_inv,
+                        remaining: remaining - 1,
+                    })
+                } else {
+                    // Create a new, independent blinding factor.
+                    reset(blinding_factor, blinding_factor_inv, e, n, rng)
+                }
+            },
+
+            None => {
+                let elem1 = try!(bigint::Elem::zero());
+                let elem2 = try!(bigint::Elem::zero());
+                reset(elem1, elem2, e, n, rng)
+            },
+        });
+
+        // Blind `x`.
+        let x =
+            try!(bigint::elem_mul_mixed(&new_contents.blinding_factor, x, n));
+
+        let x = try!(f(x));
+
+        // Unblind `x`.
+        let x =
+            try!(bigint::elem_mul_mixed(&new_contents.blinding_factor_inv, x,
+                                        n));
+
+        let _ = core::mem::replace(&mut self.0, Some(new_contents));
+
+        Ok(x)
     }
 
     #[cfg(test)]
-    pub fn counter(&self) -> u32 { unsafe { (*self.0).counter } }
-
-    pub fn as_mut_ref(&mut self) -> &mut BN_BLINDING { unsafe { &mut *self.0 } }
+    pub fn remaining(&self) -> usize {
+        match &self.0 {
+            &Some(Contents { remaining, .. }) => remaining,
+            &None => { 0 },
+        }
+    }
 }
 
-/// Needs to be kept in sync with `bn_blinding_st` in `crypto/rsa/blinding.c`.
-#[allow(non_camel_case_types)]
-#[repr(C)]
-pub struct BN_BLINDING {
-    a: *mut bigint::BIGNUM,
-    ai: *mut bigint::BIGNUM,
-    counter: u32,
-}
+fn reset(arbitrary1: bigint::Elem<N>, arbitrary2: bigint::Elem<N>,
+         e: &bigint::OddPositive, n: &bigint::Modulus<N>,
+         rng: &rand::SecureRandom) -> Result<Contents, error::Unspecified> {
+    // Use `into_elem_decoded_montgomery_encoded` to grab the underling
+    // `BIGNUM` to avoid a superfluous `malloc()` & `free()`.
+    let mut random = arbitrary1.into_elem_decoded_montgomery_encoded();
+    let mut random_inv = arbitrary2.into_elem_decoded_montgomery_encoded();
 
-extern {
-    fn GFp_BN_BLINDING_new() -> *mut BN_BLINDING;
-    fn GFp_BN_BLINDING_free(b: &mut BN_BLINDING);
-}
+    for _ in 0..32 {
+        try!(bigint::elem_randomize(&mut random, n, rng));
+        match bigint::elem_set_to_inverse_blinded(&mut random_inv, &random, n,
+                                                  rng) {
+            Ok(()) => {
+                let random = try!(bigint::elem_exp_vartime(random, e, n));
+                let random = try!(random.into_elem(n));
+                let random_inv = try!(random_inv.into_elem(n));
+                return Ok(Contents {
+                    blinding_factor: random,
+                    blinding_factor_inv: random_inv,
+                    remaining: REMAINING_MAX - 1,
+                });
+            },
+            Err(bigint::InversionError::NoInverse) => {}, // continue
+            Err(_) => { return Err(error::Unspecified); }
```

Comment:
Actually, we have:
```rust
        try!(bigint::elem_randomize(&mut random, n, rng));
        match bigint::elem_set_to_inverse_blinded(&mut random_inv, &random, n,
                                                  rng) {
```

So, we never even call `elem_set_to_inverse_blinded()`, since `elem_randomize()` will fail if the RNG only returns zeros.